### DISCUSSION TITLE
Abstract graphics

### DISCUSSION THREAD

**Discussion by dmitriitux** [joomla/joomla-cms#31815] [2. Cold Features]
Hello everyone!

Now there is a problem with the fact that joomla extensions look very different and it would be good to eliminate this by creating a standard for graphics output at the program level.
How do you like the idea to create abstract graphics, when we write not html + data in templates, but transfer data to abstract classes.
The most obvious example is Qt: https://doc.qt.io/qt-5/qtwidgets-module.html

How it might look approximate on the web:

```php
/*
 *
 * Layouts
 *
 */
 
$layout_main = new Layout();
$layout_form = new Layout();
$layout_sidebar = new Layout();

$layout_main->setDirection('horizontal');
$layout_main->add($layout_form)
       ->add($layout_sidebar);
       

/*
 *
 * Form
 *
 */
$form = new Form();
$fieldset = new Fieldset();
$group = new FormGroup()->add(new Label('My label'))
            ->add(new Input('My input', 'date'));
$fieldset->add(group);
$button = new Button('My button', 'primary');

$form->add($fieldset)
   ->add($button);

$layout_form->add($form);
```

That is, we are creating a software builder. You need to create an initial collection of basic classes such as buttons, forms, tables, etc.
What does it give in general? This gives us the fact that our extension templates no longer depend on css/js frameworks - they will be moved to providers, which can be easily changed. It is possible to maintain support for several css/js frameworks.
I have ideas on how to implement this, I am ready to implement it.

**Comment by dmitriitux** [comment]
Also, in theory, I thought about how to connect javascript with php classes

**Comment by PhocaCz** [comment]
Hi,

this is great idea.

I found, when overwriting J3 extensions to j4 that such kind of system is needed. One small example out of many: it will be great to just have some method to render tabs and let the layout class decide if this will be Bootstrap 2 tabs for J3 or Bootstrap 4 tabs for J4.

Of course, having the same output for all extensions is something what CMS definitely needs.

Jan

**Reply by bembelimen** [reply]
> One small example out of many: it will be great to just have some method to render tabs and let the layout class decide if this will be Bootstrap 2 tabs for J3 or Bootstrap 4 tabs for J4.

We already have: https://github.com/joomla/joomla-cms/blob/4.0-dev/libraries/src/HTML/Helpers/UiTab.php

In general I'm not in favor to abstract the layouts more... One of the strengths of Joomla! was that you can convert any HTML template very easy into Joomla! (without really knowing PHP).

Now with the layouts etc. it became more and more complex. For Frontend-Developer without (or small) PHP knowledge it's already really hard to create overrides or writing templates. Using such abstractions (and tbh I really see the benefit here from a programmers view) will make it nearly impossible for "normal people".

**Reply by dmitriitux** [reply]
why then people write in Qt, but here they cannot?

**Reply by dmitriitux** [reply]
if we write normal documentation, then normal developers won't have any problems.

**Reply by dmitriitux** [reply]
Well, whatever. In any case, I will create this, because I need to create final products in a uniform style, and not the trash that is happening now. I just propose to implement this into the kernel. I understand that it will complicate without proper documentation.

**Reply by dmitriitux** [reply]
@bembelimen You can delete this branch. I got an answer to my question. I will do it outside the core.

**Reply by dmitriitux** [reply]
https://github.com/joomla/joomla-cms/tree/4.0-dev/libraries/src/HTML/Helpers
you have already created it, only it is some kind of germ. You have already created abstract graphics, only stripped down and which you cannot expand normally.

The whole problem and Joomla in this. There are good ideas, but they are not developed in some embryo.

**Reply by Sulpher** [reply]
> Now with the layouts etc. it became more and more complex. For Frontend-Developer without (or small) PHP knowledge it's already really hard to create overrides or writing templates. Using such abstractions (and tbh I really see the benefit here from a programmers view) will make it nearly impossible for "normal people".

Sorry to say, but if we will focus on beginners, then Joomla needs to be simplified and returned to the times of ver.1.0.
Such an approach is a jump back... We should think about the future and how to make Joomla more flexible and even its current architecture gives us an advantage, why not make it better? 
Users with basic knowledge shouldn't pay attention to complex stuff until they get enough qualification, and as @dmitriitux said, a good tutorial can help.

**Comment by HLeithner** [comment]
@dmitriitux beside the fact that qt is a desktop library and joomla is a web cms using a browser to render webpages the idea doesn't sound bad.

But there have to be a balance between writing layouts in code like 1995 and using a completely flexible and everything can be done in anyway markup language.

Since you plan to write it anyway it would be great to see you concept/big picture.

**Reply by dmitriitux** [reply]
in fact there will be "providers" - in which the required html / javascript is unloaded
Examples providers: uikit3, bs4, bs5 - switch the provider and switch completely unloading

**Reply by HLeithner** [reply]
I have no opinion if this is good or bad for joomla at the moment, that's the reason I would like to see the complete picture.

We also have the RFC process for such things https://github.com/joomla/rfc

And I think there has to be much conceptional things to be done before someone can say, YES PERFECT that's fit in to joomla or No sorry it's a good starting point but we see this and that and here and there some problems.

**Reply by dmitriitux** [reply]
OK. I will start developing in six months. I just wanted to ask the kernel developers whether to provide this development to the kernel later.

**Reply by HLeithner** [reply]
Sounds good, if you have something to share please come back to this thread or create a new one.

**Comment by PhocaCz** [comment]
This is a very interesting topic, it looks simple, but it's very complicated. It reminds me of the endless debate Bootstrap vs Tailwind. On the one hand, something bothers us about the Bootstrap, on the other hand, we know that the Tailwind approach is also not usable everywhere and can bring a lot of problems in the future.

It is also difficult to say whether such solutions could run side by side or whether it would be possible to combine these solutions, etc., etc.

Objectives should be defined, on the basis of which the decision would then be easier.

Let's say, our goal is to have an uniform look everywhere, then the solution is different to situation where we say, we want to have the output as easy as possible, so designers have easier way to customize it.

And it's not easy at all, i think it would be different for frontend and backend.

**Comment by b2z** [comment]
If it will succeed, it will be a killer feature.
